runtime/onert/backend/train/ops/FullyConnectedLayer.cc

```diff
@@ -0,0 +1,179 @@
+
+void FullyConnectedLayer::backwardFloat32()
+{
+  // Calculate gradient for activation
+  const IPortableTensor *act_grad;
+  switch (_activation)
+  {
+    case ir::Activation::NONE:
+      act_grad = _deriv_output;
+      break;
+    case ir::Activation::RELU:
+      ReLUGrad(getShape(_output), getBuffer<float>(_output), getShape(_deriv_output),
+               getBuffer<float>(_deriv_output), getShape(_act_deriv_output.get()),
+               getBuffer<float>(_act_deriv_output.get()));
+      act_grad = _act_deriv_output.get();
+      break;
+    default:
+      throw std::runtime_error("train ConvolutionLayer: Unsupported activation type yet");
+      break;
+  }
+
+  // Initialize TransposeParams
+  nnfw::cker::TransposeParams transpose_param;
+  transpose_param.perm_count = 2;
+  transpose_param.perm[0] = 1;
+  transpose_param.perm[1] = 0;
+
+  // Initialize FullyConnectedParams
+  nnfw::cker::FullyConnectedParams op_params;
+  op_params.activation = nnfw::cker::FusedActivationFunctionType::kNone;
+
+  // Transpose and compute gradient for input
+  auto transposed_weights = _transposed_weights.get();
+  assert(transposed_weights->getShape().rank() == 2);
+  nnfw::cker::Transpose(transpose_param, getShape(_weights), getBuffer<float>(_weights),
+                        getShape(transposed_weights), getBuffer<float>(transposed_weights));
+
+  nnfw::cker::FullyConnected(op_params, getShape(act_grad), getBuffer<float>(act_grad),
+                             getShape(transposed_weights), getBuffer<float>(transposed_weights),
+                             getShape(nullptr), nullptr, getShape(_deriv_input),
+                             getBuffer<float>(_deriv_input));
+
+  // Transpose and compute gradient for weights
+  auto transposed_input = _transposed_input.get();
+  assert(transposed_input->getShape().rank() == 2);
+  nnfw::cker::Transpose(transpose_param, getShape(_input), getBuffer<float>(_input),
+                        getShape(transposed_input), getBuffer<float>(transposed_input));
+
+  auto transposed_deriv_output = _transposed_deriv_output.get();
+  assert(transposed_deriv_output->getShape().rank() == 2);
+  nnfw::cker::Transpose(transpose_param, getShape(act_grad), getBuffer<float>(act_grad),
+                        getShape(transposed_deriv_output),
+                        getBuffer<float>(transposed_deriv_output));
+
+  nnfw::cker::FullyConnected(op_params, getShape(transposed_input),
+                             getBuffer<float>(transposed_input), getShape(transposed_deriv_output),
+                             getBuffer<float>(transposed_deriv_output), getShape(nullptr), nullptr,
+                             getShape(_grad_weights), getBuffer<float>(_grad_weights));
+
+  // Compute gradient for bias
+  if (_bias)
+  {
+    assert(_grad_bias);
+    nnfw::cker::FullyConnectedBiasGrad(getShape(act_grad), getBuffer<float>(act_grad),
+                                       getShape(_grad_bias), getBuffer<float>(_grad_bias));
```

Comment:
(Q) What is benefit of implementing sum of incoming derivative in cker?

Reply:
Oh, you might want to use Eigen function for this. Am I right?

Reply:
Output is 2D and Bias is 1D, and `output.dim(1)` == `bias.dim(0)`. So The gradient of the bias is row-wise sum of the incoming gradients.

Reply:
Um, my question is about implementation location. I wonder why you implemented the details in cker . However, when I implement loss, it seems appropriate to implement it in cker to use eigen function.

Reply:
Ah..Sorry for my misunderstanding. 
If you want to use Eigen, you need to implement it in cker. As for other cases, I don't know exactly.
In fact, many of cker's code was imported from tensorflow. So, we imported the codes that had different formats in the cker. And we implemented layers in cpu backend to be able to call them in their formats.
I think it's okay to implement simple kernels wherever you feel comfortable.